smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -284,70 +266,133 @@ private void generateOperationSerializer(
         return queryBindings;
     }
 
+    private void writeHeaders(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Shape output = context.getModel().getShapeIndex().getShape(operation.getOutput().get()).get();
+
+        // Headers are always present either from the default document or the payload.
+        writer.write("let headers: any = {};");
+        writer.write("headers['Content-Type'] = $S;", bindingIndex.determineRequestContentType(
+                operation, getDocumentContentType()));
+
+        ShapeIndex index = context.getModel().getShapeIndex();
+        for (HttpBinding binding : bindingIndex.getRequestBindings(operation, Location.HEADER)) {
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            writer.openBlock("if (input.$L !== undefined) {", "}", memberName, () -> {
+                Shape target = index.getShape(binding.getMember().getTarget()).get();
+                String headerValue = getInputValue(context, binding.getLocation(), "input." + memberName,
+                        output, binding.getMember(), target);
+                writer.write("headers['$L'] = $L;", binding.getLocationName(), headerValue);
+            });
+        }
+
+        // Handle assembling prefix headers.
+        for (HttpBinding binding : bindingIndex.getRequestBindings(operation, Location.PREFIX_HEADERS)) {
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            writer.openBlock("if (input.$L !== undefined) {", "}", memberName, () -> {
+                Shape target = index.getShape(binding.getMember().getTarget()).get();
+                // Iterate through each entry in the member.
+                writer.openBlock("Objects.keys(input.$L).forEach(suffix -> {", "});", memberName, () -> {
+                    String headerValue = getInputValue(context, binding.getLocation(),
+                            "input." + memberName + "[suffix]", output, binding.getMember(), target);
+                    // Append the suffix to the defined prefix and serialize the value in to that key.
+                    writer.write("headers['$L' + suffix] = $L;", binding.getLocationName(), headerValue);
+                });
+            });
+        }
+    }
+
+    private List<HttpBinding> writeRequestBody(
```

Comment:
This method was also moved from below and includes only inconsequential updates.